readiness/controller/pkg/readiness/readiness_controller.go

```diff
@@ -0,0 +1,184 @@
+	for _, check := range readiness.Spec.Checks {
+		providers := &corev1alpha2.ReadinessProviderList{}
+		err = r.Client.List(ctxCancel, providers, &client.ListOptions{
+			FieldSelector: fields.OneTermEqualSelector("spec.checkRef", check.Name),
+		})
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		providersList.Items = append(providersList.Items, providers.Items...)
+	}
+
+	// uniqueProviders contain all the providers that satisfy at least one of the checks
+	uniqueProviders := make([]corev1alpha2.ReadinessProvider, 0)
+	uniqueProviderNames := make(map[string]bool)
+
+	for i := 0; i < len(providersList.Items); i++ {
+		if _, ok := uniqueProviderNames[providersList.Items[i].Name]; !ok {
+			uniqueProviders = append(uniqueProviders, providersList.Items[i])
+			uniqueProviderNames[providersList.Items[i].Name] = true
+		}
+	}
+
+	// allChecks contain the check names and the list of associated providers
+	allChecks := make(map[string][]int)
+	for i := 0; i < len(uniqueProviders); i++ {
+		for _, checkRef := range uniqueProviders[i].Spec.CheckRefs {
+			allChecks[checkRef] = append(allChecks[checkRef], i)
+		}
+	}
+
+	for _, check := range readiness.Spec.Checks {
+		checkStatusUpdate := corev1alpha2.CheckStatus{
+			Name:      check.Name,
+			Providers: make([]corev1alpha2.Provider, 0),
+			Ready:     false,
+		}
+
+		for _, index := range allChecks[check.Name] {
+			provider := uniqueProviders[index]
+
+			checkStatusUpdate.Ready = checkStatusUpdate.Ready || (provider.Status.State == corev1alpha2.ProviderSuccessState)
+
+			checkStatusUpdate.Providers = append(checkStatusUpdate.Providers, corev1alpha2.Provider{
+				Name:     provider.Name,
+				IsActive: provider.Status.State == corev1alpha2.ProviderSuccessState,
+			})
+		}
+		readiness.Status.CheckStatus = append(readiness.Status.CheckStatus, checkStatusUpdate)
+	}
+
+	readiness.Status.Ready = true
+
+	for _, checkStatus := range readiness.Status.CheckStatus {
+		readiness.Status.Ready = readiness.Status.Ready && checkStatus.Ready
+	}
+
+	return ctrl.Result{}, r.Client.Status().Update(ctxCancel, readiness)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ReadinessReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	err := mgr.GetFieldIndexer().IndexField(context.Background(), &corev1alpha2.ReadinessProvider{}, "spec.checkRef", func(rawObj client.Object) []string {
+		provider := rawObj.(*corev1alpha2.ReadinessProvider)
+		var indices []string
+		indices = append(indices, provider.Spec.CheckRefs...)
+		return indices
+	})
+
+	if err != nil {
+		return err
+	}
+
+	err = mgr.GetFieldIndexer().IndexField(context.Background(), &corev1alpha2.Readiness{}, "spec.checks.name", func(rawObj client.Object) []string {
+		provider := rawObj.(*corev1alpha2.Readiness)
+
+		keys := []string{}
+		for _, check := range provider.Spec.Checks {
+			keys = append(keys, check.Name)
+		}
+
+		return keys
```

Comment:
I haven't looked into how field indexers work, but doesn't this make the index on a static list of `spec.checks`. When checks are added/removed, the index isn't updated, right?

Reply:
From what I understand, [FieldIndexers](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.6/pkg/client#FieldIndexer) can perform dynamic cache updates as we are providing an IndexerFunc which will be invoked whenever any object is created/updated/deleted. Also found this [article](https://leftasexercise.com/2019/07/11/understanding-kubernetes-controllers-part-ii-object-stores-and-indexers/) that explains it.